internal/pkg/policy/output_permissions.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package policy
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/smap"
+)
+
+const (
+	DefaultOutputName         = "default"
+	OutputPermissionsProperty = "output_permissions"
+)
+
+var (
+	ErrOutputPermissionsNotFound = errors.New("output_permissions not found")
+	ErrDefaultOutputNotFound     = errors.New("default output not found")
+	ErrInvalidPermissionsFormat  = errors.New("invalid permissions format")
+)
+
+// The sample output permissions JSON from policy
+// At the moment we are only generating the key for the default output,
+// so the hashing will be on default output only
+// {
+//     "default": {
+//         "fallback": [{
+//             "names": [
+//                 "logs-*",
+//                 "metrics-*",
+//                 "traces-*",
+//                 ".logs-endpoint.diagnostic.collection-*"
+//             ],
+//             "privileges": [
+//                 "auto_configure",
+//                 "create_doc"
+//             ]
+//         }]
+//     }
+// }
+//
+// Expected to be translated into the roles descriptors format for create API key
+// https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html
+// For example:
+// {
+//     "fallback": {
+//         "index": [
+//             {
+//                 "names": [
+//                 "logs-*",
+//                 "metrics-*",
+//                 "traces-*",
+//                 ".logs-endpoint.diagnostic.collection-*"
+//                 ],
+//                 "privileges": [
+//                     "auto_configure",
+//                     "create_doc"
+//                 ]
+//             }
+//         ]
+//     }
+// }
+
+func GetRoleDescriptors(outputPermissionsRaw []byte) (hash string, roles []byte, err error) {
+	if len(outputPermissionsRaw) == 0 {
+		return
+	}
+
+	output, err := getDefaultOutputMap(outputPermissionsRaw)
+	if err != nil {
+		return
+	}
+
+	res := make(smap.Map)
+
+	for role, v := range output {
+		permissions, ok := v.([]interface{})
+		if !ok {
+			return hash, roles, ErrInvalidPermissionsFormat
+		}
+
+		idx := make([]interface{}, 0, len(permissions))
+
+		for _, permission := range permissions {
+			idx = append(idx, permission)
+		}
+
+		m := make(smap.Map)
+		m["index"] = idx
```

Comment:
I hope we can get rid of this bit as it is planned that Kibana will add the ` index` part to the output so we don't need to do the additional processing here.

Reply:
Yeah, Alejandro reached out about the upcoming change. Might add more things here then.